shared/actions/search/index.js

```diff
@@ -223,12 +228,85 @@ function* searchSuggestions<T>({payload: {actionTypeToFire, maxUsers}}: Constant
+    put(EntityAction.replaceEntity(['searchKeyToShowSearchSuggestion'], {[searchKey]: true})),
+  ])
+  yield put(Creators.finishedSearch(searchKey, ids, '', 'Keybase', true))
+}
+
+function* updateSelectedSearchResult({payload: {searchKey, id}}: Constants.UpdateSelectedSearchResult) {
+  yield put(EntityAction.replaceEntity(['searchKeyToSelectedId'], {[searchKey]: id}))
+}
+
+function* addResultsToUserInput({payload: {searchKey, searchResults}}: Constants.AddResultsToUserInput) {
+  const searchResultMap = yield select(Selectors.searchResultMapSelector)
+  const maybeUpgradedUsers = searchResults.map(u =>
+    Constants.maybeUpgradeSearchResultIdToKeybaseId(searchResultMap, u)
+  )
+  yield put.resolve(
+    EntityAction.mergeEntity(['searchKeyToUserInputItemIds'], {[searchKey]: OrderedSet(maybeUpgradedUsers)})
+  )
+  const ids = yield select(Constants.getUserInputItemIds, {searchKey})
+  yield put(Creators.userInputItemsUpdated(searchKey, ids))
+}
+
+function* removeResultsToUserInput({
+  payload: {searchKey, searchResults},
+}: Constants.RemoveResultsToUserInput) {
+  yield put.resolve(EntityAction.subtractEntity(['searchKeyToUserInputItemIds', searchKey], searchResults))
+  const ids = yield select(Constants.getUserInputItemIds, {searchKey})
+  yield put(Creators.userInputItemsUpdated(searchKey, ids))
+}
+
+function* setUserInputItems({payload: {searchKey, searchResults}}: Constants.SetUserInputItems) {
+  const ids = yield select(Constants.getUserInputItemIds, {searchKey})
+  if (!_.isEqual(ids, searchResults)) {
+    yield put.resolve(
+      EntityAction.replaceEntity(['searchKeyToUserInputItemIds'], {[searchKey]: OrderedSet(searchResults)})
+    )
+    yield put(Creators.userInputItemsUpdated(searchKey, searchResults))
+  }
+}
+
+function* clearSearchResults({payload: {searchKey}}: Constants.ClearSearchResults) {
+  yield put(EntityAction.replaceEntity(['searchKeyToResults'], {[searchKey]: null}))
+  yield put(
+    EntityAction.replaceEntity(
+      ['searchKeyToSearchResultQuery'],
+      Map({
+        [searchKey]: null,
+      })
+    )
+  )
+}
+
+function* finishedSearch({payload: {searchKey, searchResultTerm, service}}) {
+  yield put(
+    EntityAction.replaceEntity(
+      ['searchKeyToSearchResultQuery'],
+      Map({
+        [searchKey]: {text: searchResultTerm, service},
+      })
+    )
+  )
+}
+
+function* clearSearchInput({payload: {searchKey}}: Constants.UserInputItemsUpdated) {
+  const clearSearchInput = yield select(Constants.getClearSearchInput, {searchKey})
+  yield put(EntityAction.replaceEntity(['searchKeyToClearSearchInput'], {[searchKey]: clearSearchInput + 1}))
 }
 
 function* searchSaga(): SagaGenerator<any, any> {
   yield Saga.safeTakeLatest('search:search', search)
   yield Saga.safeTakeLatest('search:searchSuggestions', searchSuggestions)
+  yield Saga.safeTakeLatest('search:updateSelectedSearchResult', updateSelectedSearchResult)
```

Comment:
I could probably replace a lot of these with the pure version (which didn't exist when I first wrote this)